Implement better scaling of main table showing entries

koppor opened the pull request:
Fixes #967 so that resizing works "naturally".

Fixes https://github.com/JabRef/jabref/issues/6690

Refs https://github.com/JabRef/jabref/issues/6348

~~The only missing thing is following:~~

> ~~If there is additional space left the remaining columns could be broadend to use this space.~~

~~This is future work. Proposal: Additional button doing this "magic". Thus, the user has control when **other** than the current column should be resized.~~

~~With the current implementation, JabRef is consistent with LibreOffice Calc's behavior (and other table calculation tools).~~

~~Still open: We could not reproduce https://github.com/JabRef/jabref/issues/6690 because of https://github.com/JabRef/jabref/issues/7180~~

- [x] Change in CHANGELOG.md described (if applicable)
- [ ] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, submitted a pull request to the documentation repository.

koppor committed:
Implement better scaling of main table showing entries

Co-authored-by: Dominik Voigt <dominik.ingo.voigt@gmail.com>

tobiasdiez commented:
I think what you implemented was already possible before by unchecking "Entry table > Fit table horizontally on screen" in the preferences (except if I misunderstood what you were trying to do).

(Also I think one shouldn't take calculation programs as inspiration since they usually have the concept of infinitely many columns.)

Siedlerchr commented on src/main/java/org/jabref/gui/maintable/MainTableColumnFactory.java:
```diff
@@ -109,7 +110,16 @@ public MainTableColumnFactory(BibDatabaseContext database,
+                        columns.add(fieldColumn);
+                        if (column.getQualifier().equalsIgnoreCase(StandardField.YEAR.getName())) {
+                            // We adjust the min width and the current width, but not the max width to allow the user to enlarge this field
+                            // 75 is chosen, because of the optimal width of a four digit field
```
75 or 60?

koppor commented:
> I think what you implemented was already possible before by unchecking "Entry table > Fit table horizontally on screen" in the preferences (except if I misunderstood what you were trying to do).

Not quite:

We implemented minimum width:

https://www.loom.com/share/fc0e91e3fb4c4b178cec4de5d7e11fd7

The "Fit table horitzontally on screen" is broken:

https://www.loom.com/share/70a2a976a5164619a03ceaa150ef2cdf

* Columns can be shrunk until they are not visible anymore
* Columns on the right "jump" while moving.

We are still thinking how to proceed. In my view, a "magic alignment button" setting an "optimal" cell width and then the user can adjust the widths manually is a least WTF than the current system.

For sure, we could work on something more "intelligent". I am not sure about the user experience if we are the only tool doing that way.

> (Also I think one shouldn't take calculation programs as inspiration since they usually have the concept of infinitely many columns.)

Nevertheless, there JabRef typically has 15 columns. 

![grafik](https://user-images.githubusercontent.com/1366654/102123266-060c4680-3e47-11eb-931f-9adc61e48146.png)

The only tool I know with more than 10 columns is a spreadsheet application. Regarding the "Ewartungskonformität" (DIN 66234 Teil 8), this is the only thing raising expectations. Not sure whether other users also know spread sheet applications. However, I would guess so.

koppor commented:
Windows Explorer:

![grafik](https://user-images.githubusercontent.com/1366654/102125998-c8112180-3e4a-11eb-8414-406fce7a2687.png)

![grafik](https://user-images.githubusercontent.com/1366654/102126021-ce9f9900-3e4a-11eb-928c-5b06622cbea5.png)

Double click should set the optimum width:

![grafik](https://user-images.githubusercontent.com/1366654/102126291-2a6a2200-3e4b-11eb-8b5b-60cfe73ffd17.png)

koppor committed:
Fix number

koppor committed:
Remove resizeColumnsToFit

- Remove preferences
- Remove MainTablePreferences (and migrate left-overs into ColumnPreferences)
- Fix name of "Show dedicated file columns" (instead of "Show extra columns")

Co-authored-by: Carl Christian Snethlage <50491877+calixtus@users.noreply.github.com>
Co-authored-by: Dominik Voigt <dominik.ingo.voigt@gmail.com>

koppor committed:
Fix checkstyle

koppor commented:
DevCall decision:

Merge and work on column resizing afterwards. Probably with the magic button as in Windows Explorer.

tobiasdiez commented:
Well, the no-automatic-column-resizing strategy is the default one of JavaFX and can be enabled by unchecking the above checkbox. So no code changes required for this.

While I agree that the smart resizing policy has some flaws, it was working relatively well (for me and many users).

koppor committed:
Merge branch 'master' into fix-967

koppor committed:
Add smart resize if content fits into table width

Co-authored-by: Dominik Voigt <dominik.ingo.voigt@gmail.com>

koppor commented:
Is this really the first time, I want to delete code and @tobiasdiez wants to keep it? 😇 

@DominikVoigt and I worked hard to keep the smart resize. Here it is, freshly implemented in a (hopefully) smart way. Tested locally. Enjoyed it.

tobiasdiez commented on src/main/java/org/jabref/gui/maintable/SmartConstrainedResizePolicy.java:
```diff
@@ -22,75 +23,158 @@
+
+        for (TableColumnBase<?, ?> col : visibleLeafColumns) {
+            double share = expansionShare.get(col);
+            double newSize = col.getMinWidth() + share * remainingAvailableWidth;
```
Shouldn't this call ̀`determineNewWidth` instead (to also include the max width)?

tobiasdiez commented on src/main/java/org/jabref/gui/maintable/SmartConstrainedResizePolicy.java:
```diff
@@ -22,75 +23,158 @@
+
+            newSize = Math.min(newSize, col.getMaxWidth());
+
+            col.setPrefWidth(newSize);
```
I think this might lead to unused space on the right, when one of the columns has a higher preferred width than it's maximum (because you don't redistribute the difference among the other columns).

tobiasdiez replied:
That's why the resize method in javafx returns the unused delta:
https://github.com/openjdk/jfx/blob/53fe38bb34fc01ba298aa1a12f20c061c0cb58b2/modules/javafx.controls/src/main/java/javafx/scene/control/TableUtil.java#L363-L366

tobiasdiez commented on src/main/java/org/jabref/gui/maintable/SmartConstrainedResizePolicy.java:
```diff
@@ -22,75 +23,158 @@
+
+            col.setPrefWidth(newSize);
         }
+        return true;
```
I think you change the returned value in same cases: previously it returned false when sum of column width = table width. Not sure about the consequences.

tobiasdiez commented on src/main/java/org/jabref/gui/maintable/SmartConstrainedResizePolicy.java:
```diff
@@ -22,75 +23,158 @@
     public Boolean call(TableView.ResizeFeatures prop) {
         if (prop.getColumn() == null) {
-            return initColumnSize(prop.getTable());
+            // happens at initialization and at window resize
```
Since you below call `setPrefWidth`, after a window resize the `expansionShare` no longer reflect the actual preferred widths. 

I think the easiest solution is to use `setWidth` instead of `setPrefWidth` below: https://stackoverflow.com/a/53256244/873661

koppor replied:
We argue that the preferred size is the size the user set manually. It is difficult to distinguish whether a preferred size was set by JabRef automatically or by the user. During experiments with the functionality, it seems that maximization of the window and shrinkage and maximization and so on did not change the columns alignment. Thus, we opted to store the currently stored ratios.

For instance, that way when a user resizes the field "title" to twice the size of "author", this ratio is kept when resizing the window.

tobiasdiez replied:
Yeah, I was worried that you re-calculate the preferred width (e.g. in the initialization code), so this not necessarily need to correspond to the user's wish. But if that's not a problem, then nice.

tobiasdiez commented on src/main/java/org/jabref/gui/maintable/SmartConstrainedResizePolicy.java:
```diff
@@ -22,75 +23,158 @@
-                false,
-                getContentWidth(table) - 2,
-                visibleLeafColumns);
+        Double requiredWidth = visibleLeafColumns.stream().mapToDouble(TableColumnBase::getWidth).sum();
```
isn't this more the `currentTotalWidth`?

tobiasdiez commented on src/main/java/org/jabref/gui/maintable/SmartConstrainedResizePolicy.java:
```diff
@@ -22,75 +23,158 @@
+                        for (TableColumnBase<?, ?> col : visibleLeafColumns) {
+                            Double share = expansionShare.get(col);
+                            Double newSize;
+                            if (col.equals(prop.getColumn())) {
```
reuse userChosenColumnToResize

tobiasdiez commented on src/main/java/org/jabref/gui/maintable/SmartConstrainedResizePolicy.java:
```diff
@@ -22,75 +23,158 @@
+                                // The columns not explicitly resized by the user get the negative delta
+                                newSize = col.getWidth() - share * deltaToApply;
+                            }
+                            newSize = Math.min(newSize, col.getMaxWidth());
```
same problem as above when pref width > max width

tobiasdiez commented on src/main/java/org/jabref/gui/maintable/SmartConstrainedResizePolicy.java:
```diff
@@ -22,75 +23,158 @@
+                        double currentWidth = userChosenColumnToResize.getWidth();
+                        Double deltaToApply = newWidth - currentWidth;
+                        for (TableColumnBase<?, ?> col : visibleLeafColumns) {
+                            Double share = expansionShare.get(col);
```
Shouldn't a resize of a column not also update the expansion share directory? The user prefers to give more space to a certain column...

koppor replied:
We restore the shares now after a column was resized completely (when another column is resized)

tobiasdiez commented on src/main/java/org/jabref/gui/maintable/SmartConstrainedResizePolicy.java:
```diff
@@ -22,75 +23,158 @@
+                                newSize = newWidth;
+                            } else {
+                                // The columns not explicitly resized by the user get the negative delta
+                                newSize = col.getWidth() - share * deltaToApply;
```
I think this doesn't result in a consistent table width. Imagine having two columns, with initial preferred widths of 70/30 and a total table width of 100. Now the user extends the first column by 10, so `deltaToApply = 10`. The first column then indeed has size 80 (as expected) but the second one has `newSize = 30 - 0.3 * 10 = 27` although it should be `20`.

tobiasdiez reviewed (changes requested):
Thanks a lot for the update. Looks way better now. (And yes I like to keep useful code ;-) ).

I think I've found a few cases where the new column calculation doesn't work as expected. It would be very useful to have unit tests covering a few different scenarios. This would also give more confidence in the changes, also in the future (because I'm pretty sure we will miss some edge cases - the resize thing is really complicated sadly).
Here is a blueprint for a test: https://github.com/openjdk/jfx/blob/a128952b834fc57934c088aff1e024706dab2c5a/modules/javafx.controls/src/test/java/test/javafx/scene/control/TableViewTest.java#L1948-L1978

koppor commented on src/main/java/org/jabref/gui/maintable/MainTable.java:
```diff
@@ -137,16 +137,14 @@ public MainTable(MainTableDataModel model,
-        if (mainTablePreferences.getResizeColumnsToFit()) {
-            this.setColumnResizePolicy(new SmartConstrainedResizePolicy());
-        }
+        this.setColumnResizePolicy(new SmartConstrainedResizePolicy());
```
We removed the preference option completely - since this was a hidden feature. That should have been a visible toggle button and not hidden in the preferences.

Moreover #967 said that there should be a single behavior without the need of a configuration flat. We discussed it in length back than - and the implementors think, this is still a good thing.

koppor replied:
We added a guard to really cover both cases

tobiasdiez replied:
So you don't want to give users the option for the infinite-column behavior that you were supporting above? I don't see any harm in still having the checkbox. But since I don't want to argue again for keeping code while you want to remove it, removing is also fine with me ;-)

koppor commented on src/main/java/org/jabref/gui/maintable/MainTableColumnFactory.java:
```diff
@@ -109,7 +111,21 @@ public MainTableColumnFactory(BibDatabaseContext database,
+                        columns.add(fieldColumn);
+                        if (column.getQualifier().equalsIgnoreCase(StandardField.YEAR.getName())) {
+                            // 60 is chosen, because of the optimal width of a four digit field
+                            setExactWidth(fieldColumn, 60);
```
We thought that years of publications typically have four digits. It seems that there are other publicatoins with 5 digits? 10.000 BC? OK, we set the min width to 60.

koppor committed:
Address some comments

Co-authored-by: Dominik Voigt <dominik.ingo.voigt@gmail.com>

koppor committed:
WIP

Co-authored-by: Dominik Voigt <dominik.ingo.voigt@gmail.com>

koppor committed:
Speedup processResources

Co-authored-by: Dominik Voigt <dominik.ingo.voigt@gmail.com>

koppor committed:
Merge branch 'speedup-process-resources' into fix-967

koppor committed:
Try path without ./ at the beginning

Co-authored-by: Dominik Voigt <dominik.ingo.voigt@gmail.com>

koppor committed:
Output java error on console, too

Co-authored-by: Dominik Voigt <dominik.ingo.voigt@gmail.com>

koppor committed:
Use LOGGER instead of System.err

Co-authored-by: Dominik Voigt <dominik.ingo.voigt@gmail.com>

koppor committed:
Merge branch 'fix-output-with-wrong-jdk-version' into fix-967

koppor committed:
Merge branch 'speedup-process-resources' into fix-967

koppor committed:
Fix scroll bar still present

Co-authored-by: Dominik Voigt <dominik.ingo.voigt@gmail.com>

koppor commented:
We had some WTF moments with minwidth 10, 29. We tried with 80:

![grafik](https://user-images.githubusercontent.com/1366654/102822697-d676b480-43d9-11eb-83ad-55e7e4733561.png)

koppor committed:
MinWidth to 80

Co-authored-by: Dominik Voigt <dominik.ingo.voigt@gmail.com>

koppor committed:
Merge branch 'master' into fix-967

koppor committed:
Merge branch 'fix-967' of github.com:JabRef/jabref into fix-967

tobiasdiez commented:
I think, 10 or 15 is fine. You knows what our users put in their columns. It also makes it possible to temporarily shrink a column because it's not important for the task at hand...

koppor committed:
WIP (does not work)

Co-authored-by: Dominik Voigt <dominik.ingo.voigt@gmail.com>

koppor committed:
Merge remote-tracking branch 'origin/master' into fix-967

koppor committed:
Fix ordering in CHANGELOG.md

koppor committed:
WIP

Co-authored-by: Dominik Voigt <dominik.ingo.voigt@gmail.com>

koppor committed:
WIP - "User changed column size in a way that window can contain all the columns" still a TODO

Co-authored-by: Dominik Voigt <dominik.ingo.voigt@gmail.com>

koppor committed:
Add documentation on implementation idea

Co-authored-by: Dominik Voigt <dominik.ingo.voigt@gmail.com>

koppor committed:
Refine description (and change TableColumnBase to TableColumn)

Co-authored-by: Dominik Voigt <dominik.ingo.voigt@gmail.com>

koppor committed:
Cases I0 to I3

Co-authored-by: Dominik Voigt <dominik.ingo.voigt@gmail.com>

koppor committed:
First proposal of case names

koppor committed:
WIP: Reimplement logic according to comments

Co-authored-by: Dominik Voigt <dominik.ingo.voigt@gmail.com>

koppor committed:
Merge remote-tracking branch 'origin/master' into fix-967

koppor committed:
WIP

Co-authored-by: Dominik Voigt <dominik.ingo.voigt@gmail.com>

koppor committed:
Merge remote-tracking branch 'origin/master' into fix-967

koppor committed:
Resizing of colomms works

Co-authored-by: Dominik Voigt <dominik.ingo.voigt@gmail.com>

koppor committed:
Use EPSILON_MARGIN also for "content has fit into table before"

Co-authored-by: Dominik Voigt <dominik.ingo.voigt@gmail.com>

koppor commented:
We hit https://stackoverflow.com/a/47860355/873282. @koppor @DominikVoigt : Look at #7209. Should be easy

Expected behavior: double click: "optimal" column width. Iterate on content an choose the maximum content length as prefWidth.

koppor committed:
Merge branch 'master' into fix-967

koppor commented:
We proceed as follows:

1. @calixtus resolves merge conflicts
2. Manual test of new system
3. Merge
4. Tests will come as follow up (otherwise, this won't happen until August 2021)
5. The issue with double click on a column leading to a wrong resize will be fixed later https://github.com/JabRef/jabref/pull/7181#issuecomment-762433799
5. Right-click on table column headers will enable "fixed" columns which should improve the UX even more
6. Right-click on table column will allow to enable standard resize policy again (which has the effect that enlarging a column does not shrink others). This re-introduces the standard-behavior of windows explorer
7. Right-lock on table column will offer "reaarange all columns" (because when changing from windows explorer policy to smart resize policy could lead to bad column widths)

tobiasdiez commented on src/main/resources/log4j2.xml:
```diff
@@ -2,10 +2,13 @@
         </Console>
         <GuiAppender name="GuiLogger"/>
         <OurApplicationInsightsAppender name="applicationInsightsAppender"/>
+        <File name="FILE" fileName="jabref.log">
```
These changes here probably need to be reverted, right?

tobiasdiez commented on src/main/java/org/jabref/gui/maintable/SmartConstrainedResizePolicy.java:
```diff
@@ -1,104 +1,481 @@
+    /**
+     * SIDE EFFECT: Stores computed desired width if not present in HashMap. This leads to a constant desired width.
+     */
+    private Double getDesiredColumnWidth(TableColumn column) {
```
then rename to `computeDesiredWidth`

tobiasdiez reviewed (commented):
Ok, it would still be good if you could at least add some basic tests. 

Hopefuly this shouldn't be too hard. Create a table, with a fixed width. Then call the reize policy and verify the result.

calixtus committed:
Merge remote-tracking branch 'upstream/master' into fix-967

# Conflicts:
#	src/main/java/org/jabref/gui/maintable/MainTable.java
#	src/main/java/org/jabref/gui/preferences/table/TableTabViewModel.java

calixtus commented:
After a little bit testing, I made those observations:
- Minimal column width is too large: my year column is three times to big at minimum size.
![grafik](https://user-images.githubusercontent.com/50491877/107116299-6a112100-6872-11eb-8265-d027e895a74d.png)
- Contraintuitive behaviour: if I change the width of a column, the column border is not on the same x position as my mouse cursor is.
![grafik](https://user-images.githubusercontent.com/50491877/107117045-b90d8500-6877-11eb-9e38-72401258ca8b.png)
- After closing an reopening stored column widths are not reloaded

calixtus commented:
I fixed the merge conflicts as promised, but I believe, there are probably more cleanups to do...

koppor commented:
> After a little bit testing, I made those observations:
> 
>     * Minimal column width is too large: my year column is three times to big at minimum size.

Depends on the other columns and the other columns. We decided not to add a fixed width to the year column. We hope for a context menu to have the year column fixed. IMHO 99% the users will use the fixed width, only some non-mainstream research will use more than 4 digits for a year

>     * Contraintuitive behaviour: if I change the width of a column, the column border is not on the same x position as my mouse cursor is. The mouse cursor seems to move in-proportionally more than the column border.

We know that. At first, it also seemed to be counter-intuitive. Later, we liked it somehow (and hoped for some feedback). -- The only possibility is to change the behavior that only columns right of the current column are resized. Maybe, this is the way to go.

>     * After closing an reopening JabRef stored column widths are not reloaded

Works as designed. See https://github.com/JabRef/jabref/blob/286e7f9434c681d8b5c413a619859b4b2a62154c/src/main/java/org/jabref/gui/maintable/SmartConstrainedResizePolicy.java#L66 for a reasoning.

I repeat it here

Initial table rendering

Decision to take: a) Use last setting or b) rerender with desired widths. We opt for b), because we assume that user-configuration of desired widths is easy. This leads to I3 being inconsistent to RWS2 (there is no previous width in I3).

tobiasdiez commented:
Restoring the last column widths works currently and is not something that should be controlled by the resize policy.
My guess is that
https://github.com/JabRef/jabref/pull/7181/files#diff-535037ff480247bb2ecacf90ae5281974308e819eb2c3eed6b5cb15931a077caR122-R125
is overwriting 
https://github.com/JabRef/jabref/blob/e66f5be7d82c7a7618fe327fe958ec57a25ea598/src/main/java/org/jabref/gui/maintable/columns/MainTableColumn.java#L17-L21

calixtus commented:
Dev-Call decision: restore last state

koppor committed:
Merge branch 'main' into fix-967

Needed to add "preferences.getColumnPreferences().isDedicatedFileColumnsEnabled()))" to PersitenceVisualStateTable.java

# Conflicts:
#	docs/adr/0016-mutable-preferences-objects.md
#	src/main/java/org/jabref/gui/maintable/PersistenceVisualStateTable.java
#	src/main/java/org/jabref/gui/maintable/SmartConstrainedResizePolicy.java
#	src/main/resources/l10n/JabRef_en.properties

Siedlerchr commented:
Related https://github.com/openjdk/jfx/pull/897

koppor commented:
I didn't have the time to finish it. Moving to koppor to reduce the number of PRs here. https://github.com/koppor/jabref/pull/644
